File: Src/ButterflyNet.Parser/Strategies/MacroStrategy.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using ButterflyNet.Parser.Scopes;
namespace ButterflyNet.Parser.Strategies {
	[TokenTransformer(Token)]
	public class MacroStrategy : FunctionalStrategy {
		public const string Token = "[::";

		public MacroStrategy() {
			SetEventDelegates();
		}

		private void SetEventDelegates() {
			var startIndex = -1;

			BeforeExecute += context => {
				startIndex = context.Input.Index - (Token.Length - 1);
			};

			AfterScopeCloses += (scope, context) => {
				Debug.Assert(startIndex >= 0);

				//replace the macro declaration with its value
				var value = ((MacroScope)scope).Macro.GetValue();
				var endIndex = context.Input.Index + 1;
				context.Input.Replace(startIndex, endIndex, value);// = beginning + value + end;
				context.Input.Seek(startIndex);
				startIndex = -1;
			};
		}

		public override int Priority {
			get {
				//must be less than ModuleStrategy
				return DefaultPriority - 2;
			}
		}

		protected override Type Type { get { return ScopeTypeCache.Macro; } }

		protected override IScope CreateScope(string name, IDictionary<string, string> data, ParseContext context) {
			var macro = context.MacroFactory.Create(name);
			macro.Load(data);
			return new MacroScope(macro);
		}
	}
}

File: src/js/src/strategies/MacroStrategy.js
function MacroStrategy() {
	ModuleStrategy.$parent.call(this);
	
	this.setAsTokenTransformer("[::");
	this.priority = ParseStrategy.defaultPriority - 2;
	
	var startIndex = -1;
	
	this.beforeExecute.attach(function(context) {
		startIndex = context.input.getIndex() - 2; //token length - 1
	});
	
	this.afterScopeCloses.attach(function(scope, context) {
		var value = scope.macro.getValue();
		var endIndex = context.input.getIndex() + 1;
		
		context.input.replace(startIndex, endIndex, value);
		
		context.input.seek(startIndex - 1);
		startIndex = -1;
	});
	
	this.createScope = function(name, data, context) {
		var macro = context.macroFactory.create(name);
		macro.load(data);
		return new MacroScope(macro);
	};
}

extend(FunctionalStrategy, MacroStrategy);